File: Interview-System Design Practice/Library Information System/Java/patrons/Patron.java
package patrons;

import items.Item;
import items.LoanableItem;
import library.Library;

import java.util.ArrayList;
import java.util.Date;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Patron class.
 *
 * @author Ziang Lu
 */
public class Patron {

    /**
     * Library card number assigned to this patron.
     */
    private final int cardNum;
    /**
     * Name of this patron.
     */
    private final String name;
    /**
     * Birthday of this patron.
     */
    private final Date bday;
    /**
     * Phone number of this patron.
     */
    private long phoneNum;
    /**
     * Address of this patron.
     */
    private String addr;
    /**
     * Patron's checked out items by title.
     */
    private Map<String, Item> myItemsByTitle;

    /**
     * Constructor with parameter.
     * @param cardNum library card number assigned to the patron
     * @param name name of the patron
     * @param bday birthday of the patron
     * @param phoneNum phone number of the patron
     */
    public Patron(int cardNum, String name, Date bday, long phoneNum) {
        this.cardNum = cardNum;
        this.name = name;
        this.bday = bday;
        this.phoneNum = phoneNum;
        myItemsByTitle = new HashMap<>();
    }

    /**
     * Accessor of cardNum.
     * @return cardNum
     */
    public int getCardNum() {
        return cardNum;
    }

    /**
     * Accessor of name.
     * @return name
     */
    public String getName() {
        return name;
    }

    /**
     * Calculates the age of this patron.
     * @return age of the patron
     */
    public int getAge() {
        Date curr = new Date();
        return curr.getYear() - bday.getYear();
    }

    /**
     * Accessor of phoneNum.
     * @return phoneNum
     */
    public long getPhoneNum() {
        return phoneNum;
    }

    /**
     * Accessor of addr.
     * @return addr
     */
    public String getAddr() {
        return addr;
    }

    /**
     * Returns the checked out items by this patron as a list.
     * @return checked out items by this patron
     */
    public List<Item> getMyItems() {
        return new ArrayList<>(myItemsByTitle.values());
    }

    /**
     * Mutator of addr.
     * @param addr address to set
     */
    public void setAddr(String addr) {
        this.addr = addr;
    }

    /**
     * Requests to the given library for the item with the given title.
     * @param lib library to make the request
     * @param title title of the item to request
     */
    public void requestItem(Library lib, String title) {
        // Children (age 12 or under) can only check out 5 items at a time
        if ((getAge() <= 12) && (myItemsByTitle.values().size() >= 5)) {
            return;
        }

        Item loanable = lib.itemRequested(title);
        if (loanable == null) {
            return;
        }
        myItemsByTitle.put(title, loanable);
    }

    /**
     * Renews the given item to the given library.
     * @param lib library to apply for renew
     * @param title title of the item to renew
     */
    public void renewItem(Library lib, String title) {
        if (!myItemsByTitle.containsKey(title)) {
            return;
        }
        Item item = myItemsByTitle.get(title);
        if (!lib.itemRenewed(item)) {
            returnItem(lib, item);
        }
    }

    /**
     * Private helper method to return the given item to the given library.
     * @param lib library to return to
     * @param item item to return
     */
    private void returnItem(Library lib, Item item) {
        lib.itemReturned(item);
        myItemsByTitle.remove(item.getTitle());
    }

    /**
     * Returns the item with the given title to the given library.
     * @param lib library to return to
     * @param title title of the item to return
     */
    public void returnItem(Library lib, String title) {
        if (!myItemsByTitle.containsKey(title)) {
            return;
        }l
        Item item = myItemsByTitle.get(title);
        returnItem(lib, item);
    }

}

File: Interview-System Design Practice/Library Information System/Python/patron.py
#!usr/bin/env python3
# -*- coding: utf-8 -*-

"""
Patron module.
"""

__author__ = 'Ziang Lu'

from datetime import date
from typing import Iterable

from items import Item, LoanableItem
from library import Library


class Patron:

    def __init__(self, card_num: int, name: str, bday):
        """
        Constructor with parameter.
        :param card_num: int
        :param name: str
        :param bday: datetime object
        """
        self._card_num = card_num
        self._name = name
        self._bday = bday
        self._phone_num = 0
        self._addr = None
        self._my_items_by_title = {}

    @property
    def card_num(self) -> int:
        """
        Accessor of card_num.
        :return: int
        """
        return self._card_num

    @property
    def name(self) -> str:
        """
        Accessor of name.
        :return: str
        """
        return self._name

    @property
    def age(self) -> int:
        """
        Returns the age of this patron.
        :return: int
        """
        today = date.today()
        return today.year - self._bday.year

    @property
    def phone_num(self) -> int:
        """
        Accessor of phone_num.
        :return: int
        """
        return self._phone_num

    @property
    def addr(self) -> str:
        """
        Accessor of addr.
        :return: str
        """
        return self._addr

    @property
    def items(self) -> Iterable[Item]:
        """
        Returns the checked out items by this patron as a list.
        :return: iterable[Item]
        """
        return self._my_items_by_title.values()

    @phone_num.setter
    def phone_num(self, phone_num: int) -> None:
        """
        Mutator of phone_num.
        :param phone_num: int
        :return: None
        """
        self._phone_num = phone_num

    @addr.setter
    def addr(self, addr: str) -> None:
        """
        Mutator of addr.
        :param addr: str
        :return: None
        """
        self._addr = addr

    def request_item(self, lib: Library, title: str) -> None:
        """
        Requests to the given library for the item with the given title.
        :param lib: Library
        :param title: str
        :return: None
        """
        # Children (age 12 or under) can only check out 5 items at a time
        if self.age <= 12 and len(self._my_items_by_title.values()) >= 5:
            return

        loanable = lib.item_requested(title)
        if not loanable:
            return
        self._my_items_by_title[title] = loanable

    def renew_item(self, lib: Library, title: str) -> None:
        """
        Renews the given item to the given library.
        :param lib: Library
        :param title: str
        :return: None
        """
        if title not in self._my_items_by_title:
            return
        item = self._my_items_by_title[title]
        if not lib.item_renewed(item):
            self._return_item(lib, item)

    def _return_item(self, lib: Library, item: LoanableItem) -> None:
        """
        Private helper function to return the given item to the given library.
        :param lib: Library
        :param item: Item
        :return: None
        """
        lib.item_returned(item)
        self._my_items_by_title.pop(item.title)

    def return_item(self, lib: Library, title: str) -> None:
        """
        Returns the item with the given title to the given library.
        :param lib: Library
        :param title: str
        :return: None
        """
        if title not in self._my_items_by_title:
            return
        item = self._my_items_by_title[title]
        self._return_item(lib, item)

    def calc_overdue_fine(self) -> float:
        """
        Calculates this patron's overdue fine.
        :return: float
        """
        today = date.today()
        overdue_fine = 0.0
        for item in self._my_items_by_title.values():
            due_date = item.due_date
            if due_date < today:
                diff = today - due_date
                delayed_days = diff.days
                # The overdue fine for an item cannot be higher than the value
                # of the item.
                overdue_fine += min(delayed_days * 0.1, item.value)
        return overdue_fine
